include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h

```diff
@@ -0,0 +1,427 @@
+        int nr_points;
+
+        /** \brief 3D voxel centroid */
+        Eigen::Vector3d mean_;
+
+        /** \brief Nd voxel centroid
+         * \note Differs from \ref mean_ when color data is used
+         */
+        Eigen::VectorXf centroid;
+
+        /** \brief Voxel covariance matrix */
+        Eigen::Matrix3d cov_;
+
+        /** \brief Inverse of voxel covariance matrix */
+        Eigen::Matrix3d icov_;
+
+        /** \brief Eigen vectors of voxel covariance matrix */
+        Eigen::Matrix3d evecs_;
+
+        /** \brief Eigen values of voxel covariance matrix */
+        Eigen::Vector3d evals_;
+
+      };
+
+      struct LeafID {
+        std::string voxel_id;
+        int leaf_id;
+        bool operator < (const LeafID& rhs) const {
+          if (voxel_id < rhs.voxel_id) {
+            return true;
+          }
+          if (voxel_id > rhs.voxel_id) {
+            return false;
+          }
+          if (leaf_id < rhs.leaf_id) {
+            return true;
+          }
+          if (leaf_id > rhs.leaf_id) {
+            return false;
+          }
+          return false;
+        }
+      };
+
+      /** \brief Pointer to MultiVoxelGridCovariance leaf structure */
+      typedef Leaf* LeafPtr;
+
+      /** \brief Const pointer to MultiVoxelGridCovariance leaf structure */
+      typedef const Leaf* LeafConstPtr;
+
+      typedef std::map<LeafID, Leaf> Map;
+
+      struct VoxelGridInfo {
+        PointCloud voxel_centroids;
+        Map leaves;
+        std::vector<LeafID> leaf_indices;
+        void clear(){
+          voxel_centroids.clear();
+          leaves.clear();
+          leaf_indices.clear();
+        }
+      };
+
+      struct BoundingBox
+      {
+        Eigen::Vector4i max;
+        Eigen::Vector4i min;
+        Eigen::Vector4i div_mul;
+      };
+
+    public:
+
+      /** \brief Constructor.
+       * Sets \ref leaf_size_ to 0 and \ref searchable_ to false.
+       */
+      MultiVoxelGridCovariance () :
+        searchable_ (true),
+        min_points_per_voxel_ (6),
+        min_covar_eigvalue_mult_ (0.01),
+        leaves_ (),
+        voxel_grid_info_dict_ (),
+        voxel_centroids_leaf_indices_ (),
+        kdtree_ (),
+        voxel_grid_info_all_ ()
+      {
+        // downsample_all_data_ = false;
+        // save_leaf_layout_ = false;
+        leaf_size_.setZero ();
+        min_b_.setZero ();
+        max_b_.setZero ();
+        filter_name_ = "MultiVoxelGridCovariance";
+      }
+
+      /** \brief Initializes voxel structure.
+       * \param[in] searchable flag if voxel structure is searchable, if true then kdtree is built
+       */
+      inline void
+      setInputCloudAndFilter (const PointCloudConstPtr &cloud, const std::string cloud_id, bool searchable = false)
+      {
+        searchable_ = searchable;
+        VoxelGridInfo voxel_grid_info;
+        applyFilter ((*cloud).makeShared(), cloud_id, voxel_grid_info); // TODO koji: yabai
```

Comment:
It looks yabai copy.
I think we can avoid redundant copy by changing the argument type of the `applyFilter()`. Wouldn't that work?

`void applyFilter (const PointCloud`**Const**`Ptr &input, const std::string & cloud_id, VoxelGridInfo &voxel_grid_info) const;`
then, 
`applyFilter (cloud, cloud_id, voxel_grid_info);`

Reply:
Thanks, I completely forgot this ToDo.
It seems that your proposal looks nice. I reflected the idea here: https://github.com/tier4/ndt_omp/pull/19/commits/bcf24a826ddf1a356cb21f5d76392819bc59344f